source/agora/consensus/Validation.d

```diff
@@ -223,38 +224,225 @@ unittest
 
     Params:
         block = the block to check
         prev_height = the height of the previous block which this
                       block should point to
+        findOutput = delegate to find the referenced unspent UTXOs with
 
     Returns:
         true if the block is considered valid
 
 *******************************************************************************/
 
-public bool isValid (const ref Block block, in ulong prev_height)
-    pure nothrow @safe @nogc
+public bool isValid (const ref Block block, in ulong prev_height,
+    in Hash prev_hash, UtxoFinder findOutput) nothrow @safe
 {
+    import std.algorithm;
+
+    // special case for the genesis block
+    if (block.header.height == 0)
+        return block == getGenesisBlock();
+
     if (block.header.height != prev_height + 1)
         return false;
 
+    if (block.header.prev_block != prev_hash)
+        return false;
+
+    if (block.txs.length != Block.TxsInBlock)
+        return false;
+
+    if (block.txs.any!(tx => !tx.isValid(findOutput)))
+        return false;
+
+    Hash[] merkle_tree;
+    if (block.header.merkle_root != Block.buildMerkleTree(block.txs, merkle_tree))
+        return false;
+
     return true;
 }
 
 ///
 unittest
 {
     import agora.consensus.Genesis;
+    import std.algorithm;
+    import std.range;
+
+    // note: using array as a workaround to be able to store const Transactions
+    const(Transaction)[][Hash] tx_map;
+    scope findOutput = (Hash hash, size_t index)
+    {
+        if (auto tx = hash in tx_map)
+        {
+            if (index < (*tx).front.outputs.length)
+                return &(*tx).front.outputs[index];
+        }
+
+        return null;
+    };
+
+    auto gen_key = getGenesisKeyPair();
     auto gen_block = getGenesisBlock();
-    Block block;
-    block.header.prev_block = gen_block.header.hashFull();
-    block.header.height = gen_block.header.height + 1;
-    assert(block.isValid(gen_block.header.height));
+    assert(gen_block.isValid(gen_block.header.height, Hash.init, null));
+    auto gen_hash = gen_block.header.hashFull();
+
+    auto gen_tx = gen_block.txs[0];
+    tx_map[gen_tx.hashFull()] = [gen_tx];
+    auto txs = makeChainedTransactions(gen_key, null, 1);
+    auto block = makeNewBlock(gen_block, txs);
+
+    // height checks
+    assert(block.isValid(gen_block.header.height, gen_hash, findOutput));
 
     block.header.height = 100;
-    assert(!block.isValid(gen_block.header.height));
+    assert(!block.isValid(gen_block.header.height, gen_hash, findOutput));
+
+    block.header.height = gen_block.header.height + 1;
+    assert(block.isValid(gen_block.header.height, gen_hash, findOutput));
+
+    /// .pref_block check
+    block.header.prev_block = block.header.hashFull();
+    assert(!block.isValid(gen_block.header.height, gen_hash, findOutput));
+
+    block.header.prev_block = gen_hash;
+    assert(block.isValid(gen_block.header.height, gen_hash, findOutput));
+
+    /// .txs length check
```

Comment:
Since you are not rebuilding the merkle root it would not succeed anyway, would it ?

Reply:
Well, the length check comes before the expensive merkle root check, so I've kept it in. And I can't rebuild the merkle root because it also contains the .length check. Once we remove the transaction length limit, we can remove this check too.